server/market/market.go

```diff
@@ -3,30 +3,251 @@
+	go m.runEpochs()
+}
+
+func (m *Market) notifyNewOrder(order order.Order, action bookUpdateAction) {
+	// send to each receiver
+	for _, s := range m.orderFeeds {
+		select {
+		case <-m.ctx.Done():
+			return
+		case s <- &BookUpdateSignal{action, order}:
+		}
+	}
+}
+
+func (m *Market) runEpochs() {
+	// Startup epoch is the epoch after the epoch one second in the future.
+	dur := int64(m.epochDuration)
+	nextEpochIdx := 1 + (1+time.Now().Unix())%dur
+	nextEpoch := NewEpoch(nextEpochIdx, dur)
+	epochCycle := time.After(time.Until(nextEpoch.Start))
+
+	defer m.wg.Done()
+
+	for {
+		select {
+		case <-m.ctx.Done():
+			m.running = false
+			return
+
+		case <-epochCycle:
+			m.running = true
+			if m.epochQueue != nil {
+				go m.processEpoch(m.epochQueue)
+			}
+			m.epochQueue = nextEpoch
+
+			nextEpochIdx = m.epochQueue.Epoch + 1
+			nextEpoch = NewEpoch(nextEpochIdx, dur)
+			epochCycle = time.After(time.Until(nextEpoch.Start))
+
+		case s := <-m.orderRouter:
+			if m.epochQueue == nil /*!m.running*/ {
+				log.Debugf("Order %v received prior to market start.")
+				s.errChan <- ErrMarketNotRunning
+				continue
+			}
+
+			if ord := m.epochQueue.Orders[s.ord.ID()]; ord != nil {
+				// TODO: remove this check when we are sure the order router is behaving
+				log.Warnf("Received duplicate order %v!", ord)
+			}
+
+			// Validate the order. The order router must do it's own validation,
+			// but do a second validation before putting it on the queue.
+			if !m.validateOrder(s.ord) {
+				s.errChan <- ErrInvalidOrder
+				continue
+			}
+
+			// Insert the order into the epoch queue.
+			m.epochQueue.Orders[s.ord.ID()] = s.ord
+			s.errChan <- nil
+
+			go m.notifyNewOrder(s.ord, epochAction)
+
+			// archive the event?
+		}
+	}
+
+}
+
+func (m *Market) processEpoch(epoch *EpochQueue) {
+	// Create a slice of orders. The random order from the map range is not
+	// important since the Matcher applies a deterministic shuffling.
+	orders := make([]order.Order, 0, len(epoch.Orders))
+	for _, ord := range epoch.Orders {
+		orders = append(orders, ord)
+	}
+
+	// Perform the matching. The matcher updates the order book.
+	matches, _, failed, partial, booked, unbooked := m.matcher.Match(m.book, orders)
+	log.Debugf("Matching complete for epoch %d:"+
+		"- %d matches (%d partial fills), %d booked, %d unbooked, %d failed",
+		len(matches), len(partial), len(booked), len(unbooked), len(failed),
+	)
+
+	for _, ord := range booked {
+		m.notifyNewOrder(ord, bookAction)
+	}
+
+	for _, ord := range unbooked {
+		m.notifyNewOrder(ord, unbookAction)
+	}
+
+	go m.swapper.Negotiate(matches)
+}
+
+func (m *Market) validateOrder(order order.Order) bool {
+	return db.ValidateOrder(order, types.OrderStatusEpoch, m.marketInfo)
+}
+
+type orderUpdateSignal struct {
+	ord     order.Order
+	errChan chan error // should be buffered
+}
+
+func newOrderUpdateSignal(ord order.Order) *orderUpdateSignal {
+	return &orderUpdateSignal{ord, make(chan error, 1)}
+}
```

Comment:
Something like this could be exported and/or defined by the order router in https://github.com/decred/dcrdex/pull/65/files